blobstore/upload.go

```diff
@@ -220,33 +207,223 @@ func (s3Ctrl *S3Controller) GetUploadPresignedURL(bucket string, key string, exp
 	key := c.QueryParam("key")
 	if key == "" {
-		err := errors.New("`key` parameters are required")
-		log.Error("HandleGeneratePresignedURL: " + err.Error())
-		return c.JSON(http.StatusUnprocessableEntity, err.Error())
+		errMsg := fmt.Errorf("`key` parameters are required")
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+	bucket := c.QueryParam("bucket")
+	httpCode, err := bh.CheckUserS3WritePermission(c, bucket, key)
+	if err != nil {
+		errMsg := fmt.Errorf("error while checking for user permission: %s", err)
+		log.Error(errMsg.Error())
+		return c.JSON(httpCode, errMsg.Error())
 	}
+	//get controller for bucket
+	s3Ctrl, err := bh.GetController(bucket)
+	if err != nil {
+		errMsg := fmt.Errorf("bucket %s is not available, %s", bucket, err.Error())
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+	uploadID, err := s3Ctrl.GetMultiPartUploadID(bucket, key)
+	if err != nil {
+		errMsg := fmt.Errorf("error retrieving MultiPart Upload ID: %s", err.Error())
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusInternalServerError, errMsg.Error())
+	}
+	log.Infof("successfully generated MultiPart Upload ID for key: %s", key)
+	return c.JSON(http.StatusOK, uploadID)
+}
+
+// function that will complete a multipart upload ID when all parts are completely uploaded
+func (s3Ctrl *S3Controller) CompleteMultipartUpload(bucket string, key string, uploadID string, parts []*s3.CompletedPart) (*s3.CompleteMultipartUploadOutput, error) {
+	input := &s3.CompleteMultipartUploadInput{
+		Bucket:   aws.String(bucket),
+		Key:      aws.String(key),
+		UploadId: aws.String(uploadID),
+		MultipartUpload: &s3.CompletedMultipartUpload{
+			Parts: parts,
+		},
+	}
+	result, err := s3Ctrl.S3Svc.CompleteMultipartUpload(input)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
 
+// endpoint handler that will complete a multipart upload
+func (bh *BlobHandler) HandleCompleteMultipartUpload(c echo.Context) error {
+	key := c.QueryParam("key")
+	if key == "" {
+		errMsg := fmt.Errorf("`key` parameters are required")
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
 	bucket := c.QueryParam("bucket")
-	if bh.Config.AuthLevel > 0 {
-		claims, ok := c.Get("claims").(*auth.Claims)
-		if !ok {
-			return c.JSON(http.StatusInternalServerError, "Could not get claims from request context")
+	s3Ctrl, err := bh.GetController(bucket)
+	if err != nil {
+		errMsg := fmt.Errorf("bucket %s is not available, %s", bucket, err.Error())
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+	httpCode, err := bh.CheckUserS3WritePermission(c, bucket, key)
+	if err != nil {
+		errMsg := fmt.Errorf("error while checking for user permission: %s", err)
+		log.Error(errMsg.Error())
+		return c.JSON(httpCode, errMsg.Error())
+	}
+	type Part struct {
```

Comment:
I think types should be lower case unless you are exporting outside the package. Can you double check and update if needed